dashboard/test/controllers/api/v1/pd/regional_partner_contacts_controller_test.rb

```diff
@@ -1,44 +1,206 @@
 require 'test_helper'
 
-class Api::V1::Pd::RegionalPartnerContactsControllerTest < ::ActionController::TestCase
-  SAMPLE_FORM_DATA = {
-    first_name: 'Harry',
-    last_name: 'Potter',
-    title: 'Mr.',
-    email: 'potter@hogwarts.edu',
-    role: 'Teacher',
-    job_title: 'Defense against dark arts',
-    grade_levels: ['High School'],
-    school_type: 'public',
-    school_state: 'NY',
-    school_district_other: true,
-    school_district_name: 'Hogwarts',
-    opt_in: 'Yes',
-    notes: 'I have a question about boggarts.'
-  }
+class Api::V1::Pd::RegionalPartnerContactsControllerTest < ActionDispatch::IntegrationTest
+  test 'school type is required' do
+    refute_valid_form FOUND_DISTRICT_ONLY.merge("school-type" => "")
+  end
+
+  test 'school state is required' do
+    refute_valid_form FOUND_DISTRICT_ONLY.merge("school-state" => "")
+  end
+
+  test 'school district is required' do
+    refute_valid_form FOUND_DISTRICT_ONLY.merge("school-district" => "")
+  end
+
+  test 'can create a new regional partner contact with found district and no school' do
+    assert_valid_form FOUND_DISTRICT_ONLY
+  end
+
+  test 'can create a new regional partner contact with found district and school' do
+    assert_valid_form FOUND_DISTRICT_AND_SCHOOL
+  end
+
+  test 'can create a new regional partner contact with found district and other school' do
+    assert_valid_form FOUND_DISTRICT_OTHER_SCHOOL
+  end
+
+  test 'zip code is required for found district and other school' do
```

Comment:
I don't think the controller is the right place for all this testing. It really belongs in the model since you are testing the validator. You also don't need to have one separate case for each field that you are testing. You can try to create a regional partner contact without any info and assert that it is not valid, and that all the error messages correspond to all the missing fields.

Reply:
PTAL.

I moved my validation cases back into the model unit tests, left only happy-path cases in the controller integration test, and deduplicated my test data into factory traits.

As discussed offline, I'd prefer to keep one test case per field in these tests for a couple of reasons:

1. There's some weird dependent "required" behavior in this form.
2. This form only generates one 'generic' validation error for any missing school info.